### PR TITLE
bpdoc improvements

### PR DESCRIPTION
**Pull request by colincross**
Extract module type documentation, and support primary builder customization of factory function to use for documentation for each module type.

### CODE REVIEW

**Comment by colincross** [comment]
I'm going to refactor this.

**Comment by colincross** [comment]
Refactored.

File: bootstrap/bpdoc/reader.go [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+	if pkgPath != "" {
+		pkgPath += "/"
+	}
+	pkgPath += p[0]
+
+	return pkgPath, nil
+}
+
+func (r *Reader) ModuleType(name string, factory reflect.Value) (*ModuleType, error) {
```
Comment (gyias) [inline bootstrap/bpdoc/reader.go]: I think it'd be more reader-friendly if we move this function to bpdoc.go. Having this only return a partially filled ModuleType instance and completing it in getModuleType feel like unneeded layering that only puzzles readers.

File: bootstrap/bpdoc/reader.go [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+	}
+
+	return &ModuleType{
+		Name:    name,
+		PkgPath: pkgPath,
+		Text:    text,
+	}, nil
+}
+
+func (r *Reader) Package(pkgPath string) (*Package, error) {
```
Comment (gyias) [inline bootstrap/bpdoc/reader.go]: In a similar vein, having this function here in reader.go when the definition of the Package struct itself and all the callers of this function are in bpdoc feels unnatural to me. The way I see it, reader.go is a sort of lower-level layer that handles more primitive actions, and making it handle higher-level structures blurs the border.

File: bootstrap/bpdoc/reader.go [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+			}
+		}
+	}
+
+	return "", nil
+}
+
+// Return the PropertyStruct associated with a property struct type.  The type should be in the
+// format <package path>.<type name>
+func (r *Reader) PropertyStruct(pkgPath, name string, defaults reflect.Value) (*PropertyStruct, error) {
```
Comment (gyias) [inline bootstrap/bpdoc/reader.go]: I have a similar feeling toward PropertyStruct-related functions, but have no clear idea how to better organize them. Would be willing to rearrange them myself once getting familiar with the code base, though.

**Pull request closed without merging**